crates/uv-extract/src/stream.rs

```diff
@@ -104,9 +104,9 @@ pub async fn unzip<R: tokio::io::AsyncRead + Unpin>(
 /// Unpack the given tar archive into the destination directory.
 ///
 /// This is equivalent to `archive.unpack_in(dst)`, but it also preserves the executable bit.
-async fn untar_in<R: tokio::io::AsyncRead + Unpin, P: AsRef<Path>>(
-    archive: &mut tokio_tar::Archive<R>,
-    dst: P,
+async fn untar_in<'a>(
+    mut archive: tokio_tar::Archive<Box<dyn tokio::io::AsyncRead + Unpin + 'a>>,
```

Comment:
@BurntSushi @ibraheemdev -- How significant would you expect the penalty to be here for using dynamic dispatch?

Reply:
I suspect the cost of I/O greatly outweights the extra indirection here, though the massive line reduction suggests the calls to `poll_read` were being inlined, so there likely will be some penalty. I'm not sure how hot of a path `untar` is that it would be noticeable.

Do we need the `Box` here? `&mut dyn AsyncRead` should work, I think.

Reply:
I can't quite get `&mut dyn AsyncRead` to work:

```
error[E0308]: mismatched types
   --> crates/uv-extract/src/stream.rs:170:17
    |
170 |     Ok(untar_in(archive, target.as_ref()).await?)
    |        -------- ^^^^^^^ expected `Archive<&mut ...>`, found `Archive<&mut GzipDecoder<...>>`
    |        |
    |        arguments to this function are incorrect
    |
    = note: expected struct `tokio_tar::Archive<&mut dyn tokio::io::AsyncRead + Unpin>`
               found struct `tokio_tar::Archive<&mut async_compression::tokio::bufread::GzipDecoder<tokio::io::BufReader<R>>>`
    = help: `async_compression::tokio::bufread::GzipDecoder<tokio::io::BufReader<R>>` implements `AsyncRead` so you could box the found value and coerce it to the trait object `Box<dyn AsyncRead>`, you will have to change the expected type as well
```

Reply:
You might have to do `(&mut reader as &mut dyn ...)` before wrapping it in `Archive`.

Reply:
Success, thank you!